rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,301 @@
+
+```ts
+// `type` is one of the following keys from the DOM ValidityState interface:
+
+interface ValidityState {
+  readonly badInput: boolean;
+  readonly customError: boolean;
+  readonly patternMismatch: boolean;
+  readonly rangeOverflow: boolean;
+  readonly rangeUnderflow: boolean;
+  readonly stepMismatch: boolean;
+  readonly tooLong: boolean;
+  readonly tooShort: boolean;
+  readonly typeMismatch: boolean;
+  readonly valid: boolean;
+  readonly valueMissing: boolean;
+}
+```
+
+This means you can even show something whe the field is valid:
+
+```jsx
+<Form.ClientMessage type="valid">✅ Valid!</Form.ClientMessage>
+```
+
+#### Custom validation
+
+On top of all the built-in client-side validation types described above you can also provide your own custom validation whislt still making use of the platform's validation abilities. It uses the `customError` type present in the constraint validity API.
+
+You can pass `type="customError"` on `Form.ClientMessage` and provide an `isValid` prop with your custom validation logic.
+
+> `isValid` is called with the current value of the control and the current values of all other controls in the form.
+
+Here's a contrived example:
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.Control />
+  <Form.ClientMessage type="customError" isValid={(value, fields) => value === 'John'}>
+    Only John is allowed.
+  </Form.ClientMessage>
+</Form.Field>
+```
+
+> `isValid` can also be an `async` function (or return a promise) to perform async validation (see type below):
+
+```ts
+type CustomValidatorFn = ValidatorSyncFn | ValidatorAsyncFn;
+type ValidatorSyncFn = (value: string, fields: FormFields) => boolean;
+type ValidatorAsyncFn = (value: string, fields: FormFields) => Promise<boolean>;
+type FormFields = { [index in string]?: FormDataEntryValue };
+```
+
+#### Accessing the validity state for even more control
+
+Sometimes, you may need to access the raw validity state of a field in order to display your own icons, or interface with your own component library by passing certain props to it. You can do this by using the `Form.ValidityState` part:
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.ValidityState>
+    {(validity) => (
+      <Form.Control asChild>
+        <TextField variant="primary" state={getTextFieldState(validity)} />
+      </Form.Control>
+    )}
+  </Form.ValidityState>
+</Form.Field>
+```
+
+> Note: `Form.ValidityState` should be nested inside a `Form.Field` part and gives you access to the validity of that field.
+
+#### Server-side validation
```

Comment:
First of all, this is a component I really needed and I can't wait to try it out!!! :heart:

I have a comment regarding the Server/Client Error.

1. location-independent error message representation

> When inside a `Field` part, it will display the errors matching that field (`serverErrors[fieldName]`) When outside a `Field` part, it will display global errors that aren't tied to a specific field (`serverErrors.global`).

Based on the above, my current understanding is that a `<Form.ServerErrorMessage />` component must be located below the targeted `<Form.Field />` component in order to display an error for a specific field. However, as I've worked with Forms, there have been times when I've wanted to display an error regardless of location.

What if we treated `name` as an optional prop, and allowed the error to be expressed if there was a field associated with it?

```jsx
{ name: 'email', type: 'invalid', message: 'This is not a valid email.' }

// ok
<Form.Field name='email'>
  <Form.ServerErrorMessage />  
</Form.Field>

// also ok
<Form.Field name='email'>
  {/* ... */}
</Form.Field> // also ok

// outside of Form.Field
<div > show all errors </div>
<Form.ServerErrorMessage name='email' />
```

2. Should we separate the ServerErrorMessage component from the ClientErrorMessage component?

There are two kinds of errors that can occur in a form, client and server, but they don't seem to be different in that they should eventually be 'passed as an error' to the user.

It would make the code a bit more concise for the user if static and dynamic validation were handled by a single Error component, but I'm curious as to the reason for separating them into two.

```jsx
// for example, 
<Form.Root
  onSubmit={() => { serverside_validation }}
>
  <Form.Field name='email' isValid={(value, fields) => value === 'John'}>
    <Form.Errors>
      {/* errors: array containing both client and server errors */}
      {errors => errors.map.join}
    </Form.Errors>
  </Form.Field>
</Form.Root>
```

Reply:
Thanks for the feedback!

> 1. location-independent error message representation

Yes having an optional `name` on them to be available outside is something we have considered and may implement.
The reason we went for `Field` collocation at the moment is that it works automatically, and it's unlikely/bad ux that you'd want your errors far away from the controls. That said flexibility is always something we consider so will think about it.

> 2. Should we separate the ServerErrorMessage component from the ClientErrorMessage component?

Yeah others (https://github.com/radix-ui/primitives/pull/1977#discussion_r1115548609) have raised similar point. It's also the part I'm the last happy about so we'll probably revise that.

Thanks!